native/server/src/sqlite_brongnal.rs

```diff
@@ -0,0 +1,350 @@
+use crate::brongnal::Storage;
+use anyhow::{Context, Result};
+use ed25519_dalek::VerifyingKey;
+use prost::Message;
+use proto::parse_verifying_key;
+use rusqlite::{params, Connection};
+use std::sync::MutexGuard;
+use std::time::{SystemTime, UNIX_EPOCH};
+use std::{sync::Arc, sync::Mutex};
+use tonic::Status;
+use x25519_dalek::PublicKey as X25519PublicKey;
+
+#[derive(Debug)]
+pub struct SqliteStorage(Arc<Mutex<Connection>>);
+
+impl SqliteStorage {
+    fn connection(&self) -> tonic::Result<MutexGuard<Connection>> {
+        self.0
+            .lock()
+            .map_err(|_e| Status::internal("failed to access sqlite connection"))
+    }
+    pub fn new(connection: Connection) -> Result<Self> {
+        connection.pragma_update(None, "journal_mode", "WAL")?;
+        connection.pragma_update(None, "synchronous", "normal")?;
+        connection.pragma_update(None, "foreign_keys", "on")?;
+
+        connection
+            .execute(
+                "CREATE TABLE IF NOT EXISTS user (
+             identity STRING PRIMARY KEY,
+             key BLOB NOT NULL,
+             current_pre_key BLOB NOT NULL,
+             creation_time INTEGER NOT NULL
+         )",
+                (),
+            )
+            .context("Creating user table failed.")?;
+        connection
+            .execute(
+                "CREATE TABLE IF NOT EXISTS pre_key (
+             key BLOB PRIMARY KEY,
+             user_identity STRING NOT NULL,
+             creation_time integer NOT NULL,
+             FOREIGN KEY(user_identity) REFERENCES user(identity)
+         )",
+                (),
+            )
+            .context("Creating pre_key table failed.")?;
+        connection
+            .execute(
+                "CREATE TABLE IF NOT EXISTS message (
+             message BLOB PRIMARY KEY,
+             user_identity STRING NOT NULL,
+             creation_time integer NOT NULL,
+             FOREIGN KEY(user_identity) REFERENCES user(identity)
+         )",
+                (),
+            )
+            .context("Creating message table failed.")?;
+
+        Ok(SqliteStorage(Arc::new(Mutex::new(connection))))
+    }
+}
+
+impl Storage for SqliteStorage {
+    fn add_user(
+        &self,
+        identity: String,
+        identity_key: VerifyingKey,
+        signed_pre_key: proto::service::SignedPreKey,
+    ) -> tonic::Result<()> {
+        eprintln!("Adding user: {identity}");
```

Comment:
Why eprintln!? This isn't an error, I see the docs say progress notes but I don't think that's been consistent so far.

Reply:
From the [eprintln! docs](https://doc.rust-lang.org/std/macro.eprintln.html)
> Use eprintln! only for error and progress messages. Use println! instead for the primary output of your program.
While it is appropriate here, I have been using `println!` until I make some time to use the `log` crate.